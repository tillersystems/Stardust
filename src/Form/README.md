# Form

### Usage

```jsx
import { Form } from '@tillersystems/stardust';
```

<!-- STORY -->

### Description

A Form is used to collect, validate and submit user inputs through provided fields which can be text input, select, radio buttons, etc.  
Submit action is provided by onSubmit callback prop.

<!-- PROPS -->

### Example

```jsx
import { Form } from '@tillersystems/stardust';

render() {
  return (
      <Form onSubmit={() => {}} name="form">
        <Form.Group>
          <Form.Field label="Label">
              <Input
                type="text"
                placeholder="tape inside me"
                value=''
                onChange={() => {}}
              />
          </Form.Field>
        </Form.Group>
      </Form>
  );
}
```
