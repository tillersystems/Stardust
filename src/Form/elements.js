import styled from 'styled-components';

export const Container = styled.form`
  width: 100%;
  height: auto;
  display: flex;
  flex-direction: column;
`;
