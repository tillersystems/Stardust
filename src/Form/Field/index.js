import React, { PureComponent, Children, cloneElement } from 'react';
import PropTypes from 'prop-types';

import { Wrapper, Container, Label, ElementContainer, ElementWrapper } from './elements';

/**
 * Defines a Form Field component. This groups a field and a label.
 *
 * @return {jsx}
 */
class Field extends PureComponent {
  /** Display name. */
  static displayName = 'Form.Field';

  /** Prop types. */
  static propTypes = {
    children: PropTypes.node,
    isRow: PropTypes.bool,
    inlineLabel: PropTypes.bool,
    labelWidth: PropTypes.string,
    label: PropTypes.string,
    className: PropTypes.string,
    size: PropTypes.string,
  };

  /** Default props. */
  static defaultProps = {
    children: null,
    isRow: false,
    inlineLabel: false,
    labelWidth: '10rem',
    label: '',
    className: '',
    size: '1',
  };

  /** Internal state. */
  state = {
    // Used to save whether inner component has focus or not.
    hasFocus: false,
  };

  /**
   * Handles focus gained by the field's element.
   */
  handleFocus = () => {
    this.setState({ ...this.state, hasFocus: true });
  };

  /**
   * Handles focus lost by the field's element.
   */
  handleBlur = () => {
    this.setState({ ...this.state, hasFocus: false });
  };

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const { hasFocus } = this.state;
    const { children, label, isRow, inlineLabel, labelWidth, className, size } = this.props;

    const injected = Children.map(children, child => {
      const {
        type: { displayName },
      } = child;

      const childProps = { _onFocus: this.handleFocus, _onBlur: this.handleBlur };

      if (displayName === 'Button') {
        childProps.type = 'submit';
      }

      return cloneElement(child, childProps);
    });

    return (
      <Wrapper isRow={isRow} size={size} className={className}>
        <Container inlineLabel={inlineLabel}>
          {label !== null && !(isRow == inlineLabel && label === '') && (
            <Label labelInline={inlineLabel} width={labelWidth} hasFocus={hasFocus}>
              {label}
            </Label>
          )}
          <ElementContainer>
            <ElementWrapper>{injected}</ElementWrapper>
          </ElementContainer>
        </Container>
      </Wrapper>
    );
  }
}

export default Field;
