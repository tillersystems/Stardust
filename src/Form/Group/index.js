import React from 'react';
import PropTypes from 'prop-types';

import { Container } from './elements';

/**
 * This component is a form group
 * It will decide the organization of all fields
 */
const Group = ({ row, inlineLabels, labelsWidth, className, children }) => (
  <Container row={row} className={className}>
    {React.Children.map(children, child =>
      React.cloneElement(child, {
        isRow: row,
        inlineLabel: inlineLabels,
        labelWidth: labelsWidth,
      }),
    )}
  </Container>
);

/**
 * PropTypes validations
 */
const { node, bool, string } = PropTypes;
Group.propTypes = {
  row: bool,
  inlineLabels: bool,
  labelsWidth: string,
  className: string,
  children: node,
};

/**
 * PropTypes default values
 */
Group.defaultProps = {
  row: false,
  inlineLabels: false,
  labelsWidth: '10rem',
  className: '',
  children: null,
};

export default Group;
