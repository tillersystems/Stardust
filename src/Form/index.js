import React from 'react';
import PropTypes from 'prop-types';

import { Container } from './elements';
import Field from './Field';
import Group from './Group';

/**
 * Create a Form
 *
 * Submit action is provided by onSubmit props
 */
const Form = ({ className, children, onSubmit, name }) => (
  <Container
    className={className}
    onSubmit={e => {
      e.preventDefault();
      onSubmit();
    }}
    name={name}
  >
    {children}
  </Container>
);

Form.Field = Field;
Form.Group = Group;

/**
 * PropTypes validation
 */
const { node, string, func } = PropTypes;
Form.propTypes = {
  children: node,
  name: string,
  onSubmit: func,
  className: string,
};

/**
 * Default props
 */
Form.defaultProps = {
  children: null,
  name: '',
  onSubmit: () => {},
  className: '',
};

export default Form;
