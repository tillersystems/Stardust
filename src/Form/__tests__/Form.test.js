import React from 'react';
import { fireEvent } from '@testing-library/react';

import Form from '..';
import { Button, TextInput } from '../..';

describe('<Form />', () => {
  test('should render without a problem', () => {
    const { container } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group>
          <Form.Field label="my label">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
          <Form.Field>
            <Button primary>Hello</Button>
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render form without label without a problem', () => {
    const { container } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group>
          <Form.Field>
            <TextInput placeholder="tape inside me" />
          </Form.Field>
          <Form.Field>
            <TextInput placeholder="tape inside me" />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render row form without a problem', () => {
    const { getByTestId } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group row>
          <Form.Field label="my label 1">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
          <Form.Field hasLabel labelContent="my label 2">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const groupNode = getByTestId('form-group');
    expect(groupNode).toHaveStyleRule('flex-direction', 'row');
  });

  test('should render form with inline labels without a problem', () => {
    const { getByText } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group inlineLabels>
          <Form.Field label="my label 1">
            <TextInput />
          </Form.Field>
          <Form.Field label="my label 2">
            <TextInput />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const labelNode = getByText('my label 1');
    expect(labelNode).toHaveStyleRule('margin-right', '0.5rem');
    expect(labelNode).toHaveStyleRule('margin-left', '0');
  });

  test('should render row form with inline labels without a problem', () => {
    const { container } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group row inlineLabels>
          <Form.Field label="my label 1">
            <TextInput />
          </Form.Field>
          <Form.Field label="my label 2">
            <TextInput />
          </Form.Field>
        </Form.Group>
      </Form>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render form with labelsWidth value', () => {
    const { queryAllByText } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group inlineLabels labelsWidth="10rem">
          <Form.Field label="my label">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
          <Form.Field label="my label">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const fieldNodes = queryAllByText('my label');
    expect(fieldNodes[0]).toHaveStyleRule('width', '10rem');
    expect(fieldNodes[1]).toHaveStyleRule('width', '10rem');
  });

  test('should render form with highlighted label when has focus', () => {
    const { container, getByTestId } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group inlineLabels>
          <Form.Field label="my label 1">
            <TextInput placeholder="tape inside me" />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const fieldNode = getByTestId('input');
    fireEvent.focus(fieldNode);
    expect(container.firstChild).toMatchSnapshot();

    fireEvent.blur(fieldNode);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render form with size value', () => {
    const { queryAllByTestId } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group inlineLabels labelsWidth="10rem">
          <Form.Field label="my label 1" size="3">
            <TextInput />
          </Form.Field>
          <Form.Field label="my label 2" size="2">
            <TextInput />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const fieldNodes = queryAllByTestId('form-field');

    expect(fieldNodes[0]).toHaveStyleRule('flex', '3');
    expect(fieldNodes[1]).toHaveStyleRule('flex', '2');
  });

  test('should render form with custom width on field level', () => {
    const { getByText } = render(
      <Form onSubmit={() => {}} name="form">
        <Form.Group inlineLabels labelsWidth="10rem">
          <Form.Field label="I am customized" size="3" labelWidth="15rem">
            <TextInput />
          </Form.Field>
          <Form.Field label="I am not" size="2">
            <TextInput />
          </Form.Field>
        </Form.Group>
      </Form>,
    );

    const customFieldNode = getByText('I am customized');
    const nonCustomFieldNode = getByText('I am not');
    expect(customFieldNode).toHaveStyleRule('width', '15rem');
    expect(nonCustomFieldNode).toHaveStyleRule('width', '10rem');
  });
});
