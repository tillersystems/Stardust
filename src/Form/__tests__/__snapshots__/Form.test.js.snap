// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render form with highlighted label when has focus 1`] = `
.c0 {
  width: 100%;
  height: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: auto;
}

.c2:not(:last-child) {
  margin-bottom: 0.7rem;
  margin-right: 0;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 0;
  margin-bottom: 0.7rem;
}

.c4 {
  width: 10rem;
  min-height: 1.4rem;
  color: hsl(213,17%,20%);
  text-align: left;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.2rem;
  line-height: 1.4rem;
  margin-top: 0.7rem;
  margin-bottom: 0.7rem;
  margin-left: 0.3rem;
  margin-right: 0.5rem;
  margin-left: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 {
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1:not(:first-child) {
  margin-top: 0.7rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(207,22%,90%);
  background-color: hsl(0,100%,100%);
  border-color: hsl(200,74%,46%);
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  width: 100%;
  background-color: hsl(0,100%,100%);
  margin: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  border: none;
  padding: 0;
  text-align: left;
}

.c8::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c8::-moz-placeholder {
  font-size: 1.4rem;
}

.c8:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c8::placeholder {
  font-size: 1.4rem;
}

.c8:focus,
.c8:active {
  outline: none;
}

<form
  class="c0"
  name="form"
>
  <div
    class="c1"
    data-testid="form-group"
  >
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <label
          class="c4"
          width="10rem"
        >
          my label 1
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c8"
                data-testid="input"
                id=""
                placeholder="tape inside me"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`<Form /> should render form with highlighted label when has focus 2`] = `
<form
  class="sc-bdVaJa fMLCLO"
  name="form"
>
  <div
    class="sc-bZQynM uJaUG"
    data-testid="form-group"
  >
    <div
      class="sc-bwzfXH btkNfP"
      data-testid="form-field"
      size="1"
    >
      <div
        class="sc-htpNat bnmNtp"
      >
        <label
          class="sc-bxivhb bEmEGS"
          width="10rem"
        >
          my label 1
        </label>
        <div
          class="sc-ifAKCX futPH"
        >
          <div
            class="sc-EHOje dYYwli"
          >
            <div
              class="sc-cvbbAY jIVDhY"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="sc-jWBwVP kCOVHR"
                data-testid="input"
                id=""
                placeholder="tape inside me"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`<Form /> should render form without label without a problem 1`] = `
.c0 {
  width: 100%;
  height: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: auto;
}

.c2:not(:last-child) {
  margin-bottom: 0.7rem;
  margin-right: 0;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1:not(:first-child) {
  margin-top: 0.7rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(207,22%,90%);
  background-color: hsl(0,100%,100%);
}

.c7 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  width: 100%;
  background-color: hsl(0,100%,100%);
  margin: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  border: none;
  padding: 0;
  text-align: left;
}

.c7::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c7::-moz-placeholder {
  font-size: 1.4rem;
}

.c7:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c7::placeholder {
  font-size: 1.4rem;
}

.c7:focus,
.c7:active {
  outline: none;
}

<form
  class="c0"
  name="form"
>
  <div
    class="c1"
    data-testid="form-group"
  >
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="c6"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c7"
                data-testid="input"
                id=""
                placeholder="tape inside me"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              class="c6"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c7"
                data-testid="input"
                id=""
                placeholder="tape inside me"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`<Form /> should render row form with inline labels without a problem 1`] = `
.c0 {
  width: 100%;
  height: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: auto;
}

.c2:not(:last-child) {
  margin-bottom: 0;
  margin-right: 3.1rem;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 0;
  margin-bottom: 0.7rem;
}

.c4 {
  width: 10rem;
  min-height: 1.4rem;
  color: hsl(213,17%,20%);
  text-align: left;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.2rem;
  line-height: 1.4rem;
  margin-top: 0.7rem;
  margin-bottom: 0.7rem;
  margin-left: 0.3rem;
  margin-right: 0.5rem;
  margin-left: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 {
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1:not(:first-child) {
  margin-top: 1rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(207,22%,90%);
  background-color: hsl(0,100%,100%);
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  width: 100%;
  background-color: hsl(0,100%,100%);
  margin: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  border: none;
  padding: 0;
  text-align: left;
}

.c8::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c8::-moz-placeholder {
  font-size: 1.4rem;
}

.c8:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c8::placeholder {
  font-size: 1.4rem;
}

.c8:focus,
.c8:active {
  outline: none;
}

<form
  class="c0"
  name="form"
>
  <div
    class="c1"
    data-testid="form-group"
  >
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <label
          class="c4"
          width="10rem"
        >
          my label 1
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c8"
                data-testid="input"
                id=""
                placeholder=""
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <label
          class="c4"
          width="10rem"
        >
          my label 2
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c8"
                data-testid="input"
                id=""
                placeholder=""
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;

exports[`<Form /> should render without a problem 1`] = `
.c0 {
  width: 100%;
  height: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: auto;
}

.c2:not(:last-child) {
  margin-bottom: 0.7rem;
  margin-right: 0;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c4 {
  width: 10rem;
  min-height: 1.4rem;
  color: hsl(213,17%,20%);
  text-align: left;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.2rem;
  line-height: 1.4rem;
  margin-top: 0.7rem;
  margin-bottom: 0.7rem;
  margin-left: 0.3rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 {
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1:not(:first-child) {
  margin-top: 0.7rem;
}

.c9 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(213,17%,20%);
  background: none;
  padding: 0.75rem 1.6rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(207,22%,90%);
  background-color: hsl(0,100%,100%);
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  width: 100%;
  background-color: hsl(0,100%,100%);
  margin: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  border: none;
  padding: 0;
  text-align: left;
}

.c8::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c8::-moz-placeholder {
  font-size: 1.4rem;
}

.c8:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c8::placeholder {
  font-size: 1.4rem;
}

.c8:focus,
.c8:active {
  outline: none;
}

<form
  class="c0"
  name="form"
>
  <div
    class="c1"
    data-testid="form-group"
  >
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <label
          class="c4"
          width="10rem"
        >
          my label
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <div
              class="c7"
              data-testid="input-container"
              width="25rem"
            >
              <input
                class="c8"
                data-testid="input"
                id=""
                placeholder="tape inside me"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="c2"
      data-testid="form-field"
      size="1"
    >
      <div
        class="c3"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <button
              class="c9"
              type="submit"
            >
              Hello
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
